### PR TITLE
Add candidate notes for initializer functions.

### LINKED ISSUE #315: Add notes for initializer function candidates.
Presently no candidates are shown if an initializer function is called with incorrect arguments. The candidates should be shown, in the same was as for other types of function.

### PR DESCRIPTION
**Pull request by bearbin**
Fixes #315 

# Implementation Overview

Same as for other types of functions.

# Notes

N/A

# Checklist

- [x] Remove all TODOs, debug prints, whitespace changes
- [x] Update docs/grammar.abnf if necessary
- [x] Include tests
- [x] Add `review wanted` label, remove "WIP" in PR title

### CODE REVIEW

File: Sources/AST/Environment/Information/SpecialInformation.swift [hunk 1]
```diff
@@ -10,6 +10,17 @@ public struct SpecialInformation {
...
 
+  public var asFunctionInformation: FunctionInformation {
+    let functionDeclaration = declaration.asFunctionDeclaration
+
+    return FunctionInformation(
+      declaration: functionDeclaration,
+      typeStates: [],
+      callerCapabilities: [],
+      isMutating: false  // TODO: this isn't strictly true.
+    )
```
Comment (DJRHails) [inline Sources/AST/Environment/Information/SpecialInformation.swift]: Do you still need this?

File: Sources/SemanticAnalyzer/SemanticError.swift [hunk 2]
```diff
@@ -26,18 +26,23 @@ extension Diagnostic {
...
 
-      if candidate.callerCapabilities.count > 1 {
-        messageTail = "one of the caller capabilities in '(\(callerCapabilities))'"
+      if functionCandidate.callerCapabilities.count == 0 {
+        messageTail = ""
+      } else if functionCandidate.callerCapabilities.count > 1 {
+        messageTail = ", which requires one of the caller capabilities in '(\(callerCapabilities))'"
       } else {
-        messageTail = "the caller capability '\(callerCapabilities)'"
+        messageTail = ", which requiresthe caller capability '\(callerCapabilities)'"
```
Comment (DJRHails) [inline Sources/SemanticAnalyzer/SemanticError.swift]: requiresthe - Missing space :)

File: Sources/SemanticAnalyzer/SemanticError.swift [hunk 3]
```diff
@@ -46,9 +51,34 @@ extension Diagnostic {
...
+          messageTail = ", which requiresthe caller capability '\(callerCapabilities)'"
+        }
+
+        return Diagnostic(severity: .note, sourceLocation: candidate.declaration.sourceLocation, message: "Perhaps you meant this function\(messageTail)")
+      case .specialInformation(let candidate):
+        if candidate.declaration.isInit {
+          return Diagnostic(severity: .note, sourceLocation: candidate.declaration.sourceLocation, message: "Perhaps you meant the initializer for this struct")
+        } else {
+          // Is fallback
+          return Diagnostic(severity: .note, sourceLocation: candidate.declaration.sourceLocation, message: "Perpahs you meant this fallback function")
```
Comment (DJRHails) [inline Sources/SemanticAnalyzer/SemanticError.swift]: Perhaps?

**Comment by bearbin** [comment]
Issues you pointed out are fixed. I don't think it's necessary to put initializer arguments in the message as they are obvious from the source, in the same was as regular functions.

**Pull request merged**